### PR TITLE
`hamiltonian_expand` support for `Identity` terms

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

Since we now often switch back and forth from operations to pauli rep's and back, we can end up with terms that are identities on no wires `qml.I()`.  Individual expectation values of such a term are forbidden given may devices may not understand what to do with such a thing.  `qml.expval(qml.I())` raises a `NotImplementedError`. 

But we still want to be able take expectation values of multi-term observables with constant offsets.

**Description of the Change:**

This PR just updates `qml.transforms.hamiltonian_expand` to handle multi-term observables with constant offsets.

**Benefits:**

`hamiltonian_expand` now works with multi-term observables with constant offsets.

**Possible Drawbacks:**

This PR does not also update `sum_expand`, as it seems like that transform first needs some organizational changes.

**Related GitHub Issues:**

[sc-55654]

**Note:** Pylint was complaining about the complexity of the function so I broke it up a little bit.  Makes more lines changed, but will probably be easier to read now.

### CODE REVIEW

File: pennylane/ops/functions/equal.py [hunk 1]
```diff
@@ -281,6 +281,9 @@ def _equal_prod_and_sum(op1: CompositeOp, op2: CompositeOp, **kwargs):
     if len(op1.operands) != len(op2.operands):
         return False
 
+    if op1.pauli_rep is not None and (op1.pauli_rep == op2.pauli_rep):  # shortcut check
+        return True
+
```
Comment (albi3ro) [inline pennylane/ops/functions/equal.py:286]:
```suggestion
```

fix merge conflict.

**Review by PietropaoloFrisoni — APPROVED** [review]
Thanks @albi3ro , it looks excellent to me.

I just have a comment: in the entire source code, I don't understand the logic regarding periods at the end of statements within docstrings. In this case, in `naive_processing_fn`, there are 2 entries in `Args` with a period and one without. In `grouping_processing_fn`, no `Args` entry has a period. Anyway, it is not important, so I see no reason not to approve.

**Comment by albi3ro** [comment]
> Thanks @albi3ro , it looks excellent to me.
> 
> I just have a comment: in the entire source code, I don't understand the logic regarding periods at the end of statements within docstrings. In this case, in `naive_processing_fn`, there are 2 entries in `Args` with a period and one without. In `grouping_processing_fn`, no `Args` entry has a period. Anyway, it is not important, so I see no reason not to approve.

The logic is that I was typing quickly and not paying detailed attention.  I'll clean those docstrings up.

File: pennylane/transforms/hamiltonian_expand.py [hunk 2]
```diff
@@ -24,6 +25,118 @@
...
+
+    Returns:
+        Result: The result of the expectation value for a multi-term observable
+    """
+    dot_products = []
+    for c_group, r_group in zip(coeff_groupings, res_groupings):
+        # pylint: disable=no-member
+        if isinstance(r_group, (tuple, list, qml.numpy.builtins.SequenceBox)):
+            r_group = qml.math.stack(r_group)
+        if getattr(r_group, "shape", tuple()) == ():
```
Comment (mudit2812) [inline pennylane/transforms/hamiltonian_expand.py:45]: Why not this?

```suggestion
        if qml.math.shape(r_group) == ():
```

**Review by mudit2812 — APPROVED** [review]
Nice! Happy to see this bug fix also adds an improvement for hardware

**Comment by Qottmann** [comment]
Is it expected that despite this PR I am still getting this error:

```
dev = qml.device("default.qubit")

h = qml.dot([0.5], [qml.I()])
assert isinstance(h, qml.ops.Sum)

@qml.qnode(dev)
def qnode():
    return qml.expval(h)

qnode()
```

```
File ~/Xanadu/pennylane/pennylane/measurements/measurements.py:174, in MeasurementProcess.__init__(self, obs, wires, eigvals, id)
    172 if wires is not None:
    173     if len(wires) == 0:
--> 174         raise ValueError("Cannot set an empty list of wires.")
    175     if obs is not None:
    176         raise ValueError("Cannot set the wires if an observable is provided.")

ValueError: Cannot set an empty list of wires.
```

**Comment by Qottmann** [comment]
Is it expected that despite this PR I am still getting this error:

```
dev = qml.device("default.qubit")

h = qml.dot([0.5, 0.5], [qml.I(), qml.X(0)])
assert isinstance(h, qml.ops.Sum)

@qml.qnode(dev)
def qnode():
    return qml.expval(h)

qnode()
```

```
File ~/Xanadu/pennylane/pennylane/measurements/measurements.py:174, in MeasurementProcess.__init__(self, obs, wires, eigvals, id)
    172 if wires is not None:
    173     if len(wires) == 0:
--> 174         raise ValueError("Cannot set an empty list of wires.")
    175     if obs is not None:
    176         raise ValueError("Cannot set the wires if an observable is provided.")

ValueError: Cannot set an empty list of wires.
```

**Pull request merged**